Core/ControlMapper.cpp

```diff
@@ -88,49 +136,225 @@ void ControlMapper::SetPSPAxis(int device, char axis, float value, int stick) {
+								touchedByMapping = true;
+							}
+							float valueOther = curInput_[other];
+							float signedValue = iter->second - valueOther;
+							float ranged = (signedValue + 1.0f) * 0.5f;
+							if (direction == -1) {
+								ranged = 1.0f - ranged;
+							}
+							// NOTICE_LOG(SYSTEM, "rawValue: %f other: %f signed: %f ranged: %f", iter->second, valueOther, signedValue, ranged);
+							value += ranged;
+						} else {
+							value += iter->second;
+						}
+					} else {
+						value += iter->second;
+					}
+				} else {
+					value += iter->second;
+				}
+			}
+		}
 
-	if (pspKeys.size() && (key.flags & KEY_IS_REPEAT)) {
+		if (!touchedByMapping) {
+			continue;
+		}
+
+		// Small values from analog inputs like gamepad sticks can linger around, which is bad here because we sum
+		// up before applying deadzone etc. This means that it can be impossible to reach the min/max values with digital input!
+		// So if non-analog events clash with analog ones mapped to the same input, decay the analog input,
+		// which will quickly get things back to normal, while if it's intentional to use both at the same time for some reason,
+		// that still works, though a bit weaker. We could also zero here, but you never know who relies on such strange tricks..
+		// Note: This is an old problem, it didn't appear with the refactoring.
+		if (!changedMapping.IsAxis()) {
+			for (auto &mapping : inputMappings) {
+				if (mapping.IsAxis()) {
+					curInput_[mapping] = ReduceMagnitude(curInput_[mapping]);
+				}
+			}
+		}
+
+		value = clamp_value(value, 0.0f, 1.0f);
+
+		// Derive bools from the floats using the device's threshold.
+		// NOTE: This must be before the equality check below.
+		bool bPrevValue = virtKeys_[i] >= threshold;
+		bool bValue = value >= threshold;
+
+		if (virtKeys_[i] != value) {
+			// INFO_LOG(G3D, "vkeyanalog %s : %f", KeyMap::GetVirtKeyName(vkId), value);
+			onVKeyAnalog(changedMapping.deviceId, vkId, value);
+			virtKeys_[i] = value;
+		}
+
+		if (!bPrevValue && bValue) {
+			// INFO_LOG(G3D, "vkeyon %s", KeyMap::GetVirtKeyName(vkId));
+			onVKey(vkId, true);
+		} else if (bPrevValue && !bValue) {
+			// INFO_LOG(G3D, "vkeyoff %s", KeyMap::GetVirtKeyName(vkId));
+			onVKey(vkId, false);
+		}
+	}
+
+	return true;
+}
+
+bool ControlMapper::Key(const KeyInput &key, bool *pauseTrigger) {
+	if (key.flags & KEY_IS_REPEAT) {
 		// Claim that we handled this. Prevents volume key repeats from popping up the volume control on Android.
 		return true;
 	}
 
-	for (size_t i = 0; i < pspKeys.size(); i++) {
-		SetPSPKey(key.deviceId, pspKeys[i], key.flags);
+	std::lock_guard<std::mutex> guard(mutex_);
+
+	InputMapping mapping(key.deviceId, key.keyCode);
+
+	if (key.flags & KEY_DOWN) {
+		curInput_[mapping] = 1.0f;
+	} else if (key.flags & KEY_UP) {
+		curInput_[mapping] = 0.0f;
 	}
 
+	// TODO: See if this can be simplified further somehow.
+	bool mappingFound = KeyMap::InputMappingToPspButton(mapping, nullptr);
```

Comment:
Hm, maybe we should at least put it inside `if ((key.flags & KEY_DOWN) && key.keyCode == NKCODE_BACK) {` and put the `if (!mappingFound || key.deviceId == DEVICE_ID_DEFAULT) {` inside that too.

-[Unknown]

Reply:
Right, that'll save a little bit indeed. Fixing.